src/ctia/bundle/core.clj

```diff
@@ -340,155 +336,165 @@
+                        relationships)
+          by-type (dissoc (group-by #(ent/id->entity-type % services) all-ids) nil)
+          by-bulk-key (into {}
+                            (map (fn [[k v]]
+                                   [(bulk/bulk-key (keyword k)) v]))
+                            by-type)
+          start (System/currentTimeMillis)
+          fetched (bulk/fetch-bulk by-bulk-key identity-map services)]
+      (send-event {:service "Export bundle fetch relationships targets"
+                   :correlation-id correlation-id
+                   :time (get-epoch-second)
+                   :event-type "export-bundle"
+                   :metric (- (System/currentTimeMillis) start)})
+      (clean-bundle fetched))))
+
+(defn- scroll-with-limit
+  "Extension for store/list-records method.
+
+  First argument should be a function taking a single record and returning a value
+  used to group paginated records.
+
+  eg. `:source_ref` to fetch all records limited by `:source_ref` field.
+
+  NOTE: `params-map` must contain `:sort` instruction with the fields based on which you wish to limit responses.
+  NOTE: `params-map` must contain `:limit` instruction to set maximum amount of entities fetched by page.
+
+  NOTE: `search_after` and `offset` usage
+        The result of each invocation of `list-records` can have two main possible cases:
+
+          1. all results belong to the same record
+             in that case next call to `list-records` must use `search_after` to 'fast forward' the cursor to the next
+             record id.
+          2. results list begins with records belong to the same id but then the tail might have 1-N partitions.
+             That means the last partition in result list is incomplete and we must use `offset` to fetch next page."
+  ([f store filter-map identity-map params-map]
+   (scroll-with-limit [] [] f store filter-map identity-map params-map))
+  ([acc prev-data f store filter-map identity-map {:keys [limit] :as params-map}]
+   (let [{data :data {{:keys [offset search_after] :as next} :next} :paging}
+         (store/list-records store filter-map identity-map params-map)
+         params-map (select-keys params-map [:limit :sort])]
+     (if (nil? next)
+       (into acc (take limit (concat prev-data data)))
+       (let [[chunk & chunks] (partition-by f data)]
+         (if (empty? chunks)
+           (recur (into acc (take limit (concat prev-data chunk)))
+                  [] f store filter-map identity-map
+                  (assoc params-map :search_after search_after))
+           (recur (reduce into acc (cons (take limit (concat prev-data chunk)) (butlast chunks)))
+                  (last chunks) f store filter-map identity-map
+                  (assoc params-map :offset offset))))))))
+
+(defn relationships-filter [records related-to source-type target-type]
+  {:query (cond->> [(format "%s:(%s)" (name related-to) (string/join " OR " (map #(format "\"%s\"" %) (map :id records))))]
```

Comment:
I intended to gain as much control over query building as possible. To be sure that the fastest filter appears first. I can test the difference as well, but then the performance will depend on implementation details and might degrade in case of unrelated changes in the lib.

To be sure I understand you. Your suggestion is to add `:source_ref [id1 ...]` alongside with `:query`?

Reply:
yes. 
However I share your concern about ensuring the order.
With filter + query I confirm that we have a low control, I just know that filters order matters in ES5 (I do not know for ES7 indeed) and that we ensure `query` is the last filter in a very distant implementation detail: https://github.com/threatgrid/ductile/blob/master/src/ductile/query.clj#L64.
In any case, the filtering on the ids with a regex is a very old hack and you finish to convince me to create a task to change it:
https://github.com/advthreat/iroh/issues/6528
So keep your implementation but check that we cannot hit a lucene query limit. I did not find anything, but it's better to double check.

Reply:
[I know only one limit](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#index-max-terms-count)
the default number is 65536. I doubt we ever want to export that many in one batch :)